source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRRichTextBlockRenderer.mm

```diff
@@ -0,0 +1,167 @@
+{
+    static ACRRichTextBlockRenderer *singletonInstance = [[self alloc] init];
+    return singletonInstance;
+}
+
++ (ACRCardElementType)elemType
+{
+    return ACRRichTextBlock;
+}
+
+- (UIView *)render:(UIView<ACRIContentHoldingView> *)viewGroup
+          rootView:(ACRView *)rootView
+            inputs:(NSMutableArray *)inputs
+   baseCardElement:(ACOBaseCardElement *)acoElem
+        hostConfig:(ACOHostConfig *)acoConfig;
+{
+    std::shared_ptr<HostConfig> config = [acoConfig getHostConfig];
+    std::shared_ptr<BaseCardElement> elem = [acoElem element];
+    std::shared_ptr<RichTextBlock> rTxtBlck = std::dynamic_pointer_cast<RichTextBlock>(elem);
+    ACRUILabel *lab = [[ACRUILabel alloc] initWithFrame:CGRectMake(0,0,viewGroup.frame.size.width, 0)];
+    lab.backgroundColor = [UIColor clearColor];
+    lab.style = [viewGroup style];
+    lab.editable = NO;
+    lab.textContainer.lineFragmentPadding = 0;
+    lab.textContainerInset = UIEdgeInsetsZero;
+    lab.layoutManager.usesFontLeading = false;
+
+    NSMutableAttributedString *content = [[NSMutableAttributedString alloc] init];
+    if(rootView){
+        NSMutableDictionary *textMap = [rootView getTextMap];
+
+        for(const auto &paragraph : rTxtBlck->GetParagraphs()) {
+            for(const auto &inlineText : paragraph->GetInlines()) {
+                std::shared_ptr<TextRun> textRun = std::static_pointer_cast<TextRun>(inlineText);
+                if(textRun) {
+                    NSNumber *number = [NSNumber numberWithUnsignedLongLong:(unsigned long long)textRun.get()];
+                    NSString *key = [number stringValue];
+                    NSDictionary* data = textMap[key];
+                    NSData *htmlData = data[@"html"];
+                    NSDictionary *options = data[@"options"];
+                    NSDictionary *descriptor = data[@"descriptor"];
+                    NSString *text = data[@"nonhtml"];
+
+                    NSMutableAttributedString *textRunContent = nil;
+                    // Initializing NSMutableAttributedString for HTML rendering is very slow
+                    if(htmlData) {
+                        textRunContent = [[NSMutableAttributedString alloc] initWithData:htmlData options:options documentAttributes:nil error:nil];
+                        // Drop newline char
+                        [textRunContent deleteCharactersInRange:NSMakeRange([textRunContent length] -1, 1)];
+                        lab.selectable = YES;
+                        lab.dataDetectorTypes = UIDataDetectorTypeLink;
+                        lab.userInteractionEnabled = YES;
+                    } else {
+                        textRunContent = [[NSMutableAttributedString alloc] initWithString:text attributes:descriptor];
+                        [textRunContent deleteCharactersInRange:NSMakeRange(0, 3)];
+                        [textRunContent deleteCharactersInRange:NSMakeRange([textRunContent length] -4, 4)];
```

Comment:
Can you add a comment explaining what this is doing?

Reply:
sure; thanks

---
In reply to: [272271678](https://github.com/Microsoft/AdaptiveCards/pull/2613#discussion_r272271678) [](ancestors = 272271678)